swarm/src/lib.rs

```diff
@@ -1731,3 +1777,134 @@ mod tests {
         }))
     }
 }
+
+pub mod dial_opts {
+    use libp2p_core::{Multiaddr, PeerId};
+
+    #[derive(Debug)]
+    pub enum DialOpts {
+        // TODO:  How about folding these structs into the enum variants?
+        WithPeerId(WithPeerId),
+        WithPeerIdWithAddresses(WithPeerIdWithAddresses),
+        WithoutPeerIdWithAddress(WithoutPeerIdWithAddress),
+    }
+
+    impl DialOpts {
+        pub fn peer_id(peer_id: PeerId) -> WithPeerId {
+            WithPeerId {
+                peer_id,
+                condition: Default::default(),
+            }
+        }
+
+        pub fn unknown_peer_id() -> WithoutPeerId {
+            WithoutPeerId {}
+        }
```

Comment:
```suggestion
        pub fn address(address: Multiaddr) -> WithoutPeerIdWithAddress {
            WithoutPeerIdWithAddress { address }
        }
```
I do like the consistent syntax of `DialOpts::peer_id(..).addresses()` and `DialOpts::unknown_peer_id(..).address(..)`.
But this similar syntax suggests that `WithPeerId` and `WithoutPeerId` are alike, which is not the case because `WithPeerId` can be a "final" state from which you can call `build()`, whereas  `WithoutPeerId` is only intermediate. Not sure about this, but I feel like it could end up confusing the user, since even though it follows the same naming convention, `WithoutPeerId` is not a valid `DialOpt`.

Optionally if we follow the above suggestion, we could keep it consist by also adding
```rust
impl WithoutPeerIdWithAddress {
        pub fn peer_id(self, peer_id: PeerId) -> WithPeerIdWithAddresses {
            WithPeerIdWithAddresses {
                peer_id,
                condition: Default::default(),,
                addresses: vec![self.address],
                extend_addresses_through_behaviour: false,
            }
        }
    }
```
Though this may be a bit confusion because without a peer id you only have one address, whereas with an peer id there can be multiple.

Reply:
Say we allowed `DialOpts::address(..)` instead of `DialOpts::unknown_peer_id(..).address(..)`, would that not confuse a user into thinking that they can then later on add a peer id, i.e. that they can do `DialOpts::address(..).peer_id(..)`?

> The only problem with this is that without a peer id you only have one address, whereas with an peer id there can be multiple.

Yes, this is something that I would like to encode at the type level, thus the differentiation between the two. I am not sure how to model your suggestion without loosing the type safety.

Reply:
> Say we allowed DialOpts::address(..) instead of DialOpts::unknown_peer_id(..).address(..), would that not confuse a user into thinking that they can then later on add a peer id, i.e. that they can do DialOpts::address(..).peer_id(..)?

I agree. But with reference to the above discussion: the pattern currently already expects a fixed order in which the parameters are specified, e.g. as written above: in case of a known peer id you have to specify the `condition` before specifying the `addresses`. Isn't that kind of the same issue?

> Yes, this is something that I would like to encode at the type level, thus the differentiation between the two. I am not sure how to model your suggestion without loosing the type safety.

You are right, the type safety should be kept; `DialOpts::address(..).peer_id(..)` should not be possible.